### ISSUE TITLE
Stacked Charts GEF editor - need drop target to add scatter sets

### ISSUE THREAD

**Issue by IanMayo** [debrief/limpet#358]
@dinkoivanov - we can now add scatter sets to our stacked charts.

But, users need to be able to edit them once in the editor.  We could also improve the UI if we had a "landing pad" for scatter data.  This landing pad will mean I can refactor Debrief's drag/drop support for loading data into Stacked Charts - it's quite clumsy at the moment (since stacked charts doesn't know if the user wants to view the data values like a normal collection, or to view the timings as a scatter set).

Here is a current screenshot:
![selection_002](https://cloud.githubusercontent.com/assets/1108513/16522866/348d824c-3f98-11e6-928a-192cdfdf9393.png)

I guess this list of datasets could appear in the green area as a horizontal list, or in the pink area as vertical labels (as we do for datasets).

Here is a potential layout:
![untitled drawing](https://cloud.githubusercontent.com/assets/1108513/16522972/c357a00c-3f98-11e6-876d-079c1ed8be03.png)

**Comment by dinkoivanov** [comment]
@IanMayo I'm OK with the green outline. Note that current idea is chart title to be centered, so there will be even smaller space for the scatter sets. It might become to crowded, unless you don't expect more than 1, perhaps max 2 such sets for a chart. Regarding the layout, I think we could optimize it (for space saving) so that the "master" icon is only present if the "landing pad" is empty. Also I'm still convinced that the style icon can be dropped and the style properties can be nested under dataset properties.
If you do intend to have more that 2 scatter sets for a chart I propose the following placement for the landing pad (in blue):
![image2](https://cloud.githubusercontent.com/assets/4459598/16541487/12249902-408d-11e6-8beb-f9a71e8a0f3c.png)
It will eat some vertical space though. What do you think?

**Comment by IanMayo** [comment]
the blue area is just fine.  That's where the time (independent) axis is, and the scatter data is shown on that axis anyway.  

Good thinking.

**Comment by dinkoivanov** [comment]
Here's a screenshot of the currently implemented functionality:

![image](https://cloud.githubusercontent.com/assets/4459598/17083873/1e152fe4-51b6-11e6-994b-e5507f904e1e.png)

Open points:
- [x] Handle Add - Right now Scatter set is added when you drop certain type of collection over the Chart. We want to change that, so it is only added when you drop it over the "scatter set landing area". In addition if user tries to drop the same scatter collection over multiple charts then we need to create it once in the Shared Axis, and reuse it in the multiple charts (AppearsIn relation)
- [x] Handle Delete - When user deletes a scatter set, then it won't be used in the current chart, but will remain in the shared axis of the ChartSet object. If the scatter set user is trying to delete is only used within the current chart and nowhere else then we can delete the scatter set from the shared axis as well
- [x] Handle moving a Scatter set from one chart to another via drag & drop
- [ ] Styling object is not available right now for Scatter set objects. This needs to be exposed (via corresponding EditPart + Figure) once supported on metamodel level
- [x] The standard icon for object is used from Eclipse, so we need to incorporate appropriate icons when available
- [x] We need to have _hover_ effect as data is dragged over the ScatterSet landing pad.

**Comment by IanMayo** [comment]
```
Styling object is not available right now for Scatter set objects. This needs
to be exposed (via corresponding EditPart + Figure) once supported on 
metamodel level
```

I don't know if it's the styling you're looking for, but the metamodel does include some styling in a parent object:
![scatterstyling](https://cloud.githubusercontent.com/assets/1108513/17141055/a5da5ff2-5342-11e6-8b3c-85b8d703689b.png)

Note: now that @theanuradha has resolved how to show core and styling properties, we may not need the "Style" UI object.

**Standard icon** - have requested new one.

**Comment by IanMayo** [comment]
Waiting for resolution of #381

**Comment by IanMayo** [comment]
Hi there @dinkoivanov ,
have refactored the production of ScatterSets & Datasets.  There are now two OSGI Extensions, and two Interfaces - one for adapters that produce ScatterSets and the other which produces Datasets.

I've got stumped with the GEF elements.  We can drop the Duration time-stamps onto the ScatterSets container, but I can't get it to highlight during the drag operation.

I also got stumped on the associated edit/delete processing.  I'm ok with the EMF list manipulation, I just couldn't glue the bits together correctly.

Would you mind having a go?

Oh, and I've got a new ScatterSet icon on it's way - hopefully tomorrow.

Cheers,
Ian

**Comment by dinkoivanov** [comment]
Hi @IanMayo,
Could you help me get started with this issue? Here is what I do:
1. Import the sample_data project from Github (do copy in the workspace to preserve the Git version)
2. Open the overview.lap file
3. Right click LongContact and do Open in Analysis view to open stacked charts
4. Click "Edit" button on the stacked charts toolbar.
5. Right click the overview.lap editor and click - Administration > Generate Large dataset
6. I get to this screen:
![capture](https://cloud.githubusercontent.com/assets/4459598/17276969/3b8a8080-5740-11e6-9ea1-5ce2c5d06673.PNG)

You mention :

> We can drop the Duration time-stamps onto the ScatterSets container

How can I achieve that? What should be the result in the view/edit mode?

Regards,
Dinko

**Comment by IanMayo** [comment]
Aaah, ok.  I don't think the "Generate large dataset" covers the same time periods as the "Long contact".

So, do these steps:
- close your Stacked Charts object
- Select "Speed Three (longer)" and "Sum of Speed One, Speed Two Time".
- Right-click and select "Analysis  / Show in stacked charts"
- Switch to edit mode
- Drag "Time stamps (early) onto the Scatter Sets landing pad.
- Now resize the Stacked Charts view a tiny amount, to trigger a redraw (yes, we need to fix this)
- You will see a few extra vertical lines

![selection_002](https://cloud.githubusercontent.com/assets/1108513/17277436/85b1f9be-573a-11e6-9b15-0551cc8305e6.png)

If you switch to View Mode, then back to Edit Mode the new scatter-set will be shown in the placeholder (yes, we need to fix this).

Hopefully these steps will get you far enough.

**Comment by dinkoivanov** [comment]
@IanMayo editing is pretty much done. Add/Delete/Move is implemented with proper Undo support.

**Comment by IanMayo** [comment]
Nice work, thanks Dinko. Look forward to having a play,  but maybe not Thursday.

**Comment by IanMayo** [comment]
Fixed in #385

**Comment by IanMayo** [comment]
@dinkoivanov - having a problem.

In this implementation we add a listener to track data changes.  But, we get an NPE after removing a chart.

Steps to reproduce:
- open **data.lap**
- select _speed three (longer)_, _sum of speed one time, speed two time_, right-click and **Analysis**   - **Show in stacked charts**
- switch to edit view
- add a new chart
- add _Product of Speed One Time, ..._ to the new chart
- delete the new chart, a NPE will be thrown.

```
!ENTRY org.eclipse.ui 4 0 2016-08-11 13:14:01.358
!MESSAGE Unhandled event loop exception
!STACK 0
java.lang.NullPointerException
    at info.limpet.stackedcharts.ui.editor.parts.ChartEditPart.getSharedAxis(ChartEditPart.java:80)
    at info.limpet.stackedcharts.ui.editor.parts.ChartEditPart.deactivate(ChartEditPart.java:74)
    at org.eclipse.gef.editparts.AbstractEditPart.deactivate(AbstractEditPart.java:293)
    at org.eclipse.gef.editparts.AbstractGraphicalEditPart.deactivate(AbstractGraphicalEditPart.java:354)
    at org.eclipse.gef.editparts.AbstractEditPart.removeChild(AbstractEditPart.java:879)
    at org.eclipse.gef.editparts.AbstractEditPart.refreshChildren(AbstractEditPart.java:793)
    at info.limpet.stackedcharts.ui.editor.parts.ChartSetEditPart.access$1(ChartSetEditPart.java:1)
    at info.limpet.stackedcharts.ui.editor.parts.ChartSetEditPart$ChartSetAdapter.notifyChanged(ChartSetEditPart.java:125)
    at org.eclipse.emf.common.notify.impl.BasicNotifierImpl.eNotify(BasicNotifierImpl.java:374)
    at org.eclipse.emf.common.notify.impl.NotificationImpl.dispatch(NotificationImpl.java:1027)
    at org.eclipse.emf.common.notify.impl.NotificationImpl.dispatch(NotificationImpl.java:1032)
    at org.eclipse.emf.common.notify.impl.NotifyingListImpl.remove(NotifyingListImpl.java:718)
    at org.eclipse.emf.common.util.AbstractEList.remove(AbstractEList.java:460)
    at info.limpet.stackedcharts.ui.editor.commands.DeleteChartCommand.execute(DeleteChartCommand.java:24)
    at org.eclipse.gef.commands.CommandStack.execute(CommandStack.java:199)
```

**Comment by IanMayo** [comment]
Fixed in #388
